core/trino-main/src/main/java/io/trino/execution/scheduler/HashDistributionSplitAssigner.java

```diff
@@ -0,0 +1,236 @@
+                for (Integer taskPartition : createdTaskPartitions) {
+                    assignment.sealPartition(taskPartition);
+                }
+                assignment.setNoMorePartitions();
+                replicatedSplits.clear();
+            }
+        }
+
+        return assignment.build();
+    }
+
+    @Override
+    public AssignmentResult finish()
+    {
+        checkState(!createdTaskPartitions.isEmpty(), "createdTaskPartitions is not expected to be empty");
+        return AssignmentResult.builder().build();
+    }
+
+    private static Map<Integer, TaskPartition> createOutputPartitionToTaskPartition(
+            FaultTolerantPartitioningScheme sourcePartitioningScheme,
+            Set<PlanNodeId> partitionedSources,
+            Map<PlanNodeId, OutputDataSizeEstimate> outputDataSizeEstimates,
+            boolean preserveOutputPartitioning,
+            long targetPartitionSizeInBytes)
+    {
+        int partitionCount = sourcePartitioningScheme.getPartitionCount();
+        if (sourcePartitioningScheme.isExplicitPartitionToNodeMappingPresent() ||
+                partitionedSources.isEmpty() ||
+                !outputDataSizeEstimates.keySet().containsAll(partitionedSources) ||
+                preserveOutputPartitioning) {
+            // if bucket scheme is set explicitly or if estimates are missing create one task partition per output partition
+            return IntStream.range(0, partitionCount)
+                    .boxed()
+                    .collect(toImmutableMap(Function.identity(), (key) -> new TaskPartition()));
+        }
+
+        List<OutputDataSizeEstimate> partitionedSourcesEstimates = outputDataSizeEstimates.entrySet().stream()
+                .filter(entry -> partitionedSources.contains(entry.getKey()))
+                .map(Map.Entry::getValue)
+                .collect(toImmutableList());
+        OutputDataSizeEstimate mergedEstimate = OutputDataSizeEstimate.merge(partitionedSourcesEstimates);
+        ImmutableMap.Builder<Integer, TaskPartition> result = ImmutableMap.builder();
+        PriorityQueue<PartitionAssignment> assignments = new PriorityQueue<>();
+        assignments.add(new PartitionAssignment(new TaskPartition(), 0));
+        for (int outputPartitionId = 0; outputPartitionId < partitionCount; outputPartitionId++) {
+            long outputPartitionSize = mergedEstimate.getPartitionSizeInBytes(outputPartitionId);
+            if (assignments.peek().assignedDataSizeInBytes() + outputPartitionSize > targetPartitionSizeInBytes
+                    && assignments.size() < partitionCount) {
+                assignments.add(new PartitionAssignment(new TaskPartition(), 0));
+            }
+            PartitionAssignment assignment = assignments.poll();
+            result.put(outputPartitionId, assignment.taskPartition());
+            assignments.add(new PartitionAssignment(assignment.taskPartition(), assignment.assignedDataSizeInBytes() + outputPartitionSize));
+        }
```

Comment:
When `HashDistributionSplitAssigner` is created an input size esitmate is provided (see `Map<PlanNodeId, OutputDataSizeEstimate> outputDataSizeEstimates`). Based on the estimates provided the `HashDistributionSplitAssigner` assigns output partitions to tasks (to avoid small tasks). In the previous version (with a full barrier) it was done based on the information obtained from `ExchangeSourceHandle`. However if speculative execution is allowed this must be done based on "estimates" as `ExchangeSourceHandle`s may not yet be available.

Reply:
My question is more like, if we have fixed number of partitions, shouldn't we simply try to distribute data evenly among the assignments? What's the sense of trying to respect `targetPartitionSizeInBytes`?

Reply:
It could work but you need same input size statics still. Otherwise you do not know how many partitions you should group together, to be handled by single task.

Reply:
@losipiuk : in the final result there is no information around size stats. It's a mapping from `outputPartitionId` to `TaskPartition`s. So I'm just more confused now...

Reply:
The `targetPartitionSizeInBytes` is needed to avoid creating tiny partitions. For example if the total data size is only 1GB it should be enough to create a single partition mapping all the output partitions to a single task.

Reply:
What I see if we output the map with partitionIds to a `new TaskPartition()`, so I'm not sure if we are achieving the above purposes stated above. Am I missing something obvious?

I have added the following to code to print out the returned result:
```
Map<Integer, TaskPartition> resultToBePrinted = result.buildOrThrow();
        resultToBePrinted.forEach((partitionId, taskPartition) -> {
            System.out.println(partitionId);
            if (taskPartition.isIdAssigned()) {
                System.out.println(taskPartition.getId());
            }
            else {
                System.out.println("Not assigned");
            }
        });
```
I'm seeing:
```
0
Not assigned
1
Not assigned
2
Not assigned
```

So not sure this code piece has any effect at all.

Reply:
Oh, i see.

`TaskPartition` is a placeholder. Basically what this algorithm does it groups certain output partitions to be processed by certain task partitions.

For example:

- Output partitions 1,2,3 must be processed by a separate task
- Output partitions 4,5 must be processed by a different task
- Output partitions 6,7 must be processed by a yet another task

However we are trying to avoid assigning a certain numeric id to a task at this step. The problem is that the output data size is only an esimate, and in reallity not all the tasks may have some data to process.

For example when reading a bucketed table what we know is that there are 1000 buckets. So we assign 1000 task partitions one for each bucket. But then it is possible that data is missing for a certain bucket what would create a confusing hole in task ids (you may endup with tasks 1.0.0, 1.1.0, 1.5.0 missing the 1.2.0. 1.3.0, 1.4.0 tasks). Assigning numeric ids lazily allows to avoid such gaps.